mesonbuild/cargo/builder.py

```diff
@@ -0,0 +1,284 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright © 2022-2023 Intel Corporation
+
+"""Provides helpers for building AST
+
+This is meant to make building Meson AST from foreign (largely declarative)
+build descriptions easier.
+"""
+
+from __future__ import annotations
+import builtins
+import dataclasses
+import typing as T
+
+from .. import mparser
+
+
+def _token(tid: str, filename: str, value: mparser.TV_TokenTypes) -> mparser.Token[mparser.TV_TokenTypes]:
+    """Create a Token object, but with the line numbers stubbed out.
+
+    :param tid: the token id (such as string, number, etc)
+    :param filename: the filename that the token was generated from
+    :param value: the value of the token
+    :return: A Token object
+    """
+    return mparser.Token(tid, filename, -1, -1, -1, (-1, -1), value)
+
+
+def string(value: str, filename: str) -> mparser.StringNode:
+    """Build A StringNode
+
+    :param value: the value of the string
+    :param filename: the file that the value came from
+    :return: A StringNode
+    """
+    return mparser.StringNode(_token('string', filename, value))
+
+
+def number(value: int, filename: str) -> mparser.NumberNode:
+    """Build A NumberNode
+
+    :param value: the value of the number
+    :param filename: the file that the value came from
+    :return: A NumberNode
+    """
+    return mparser.NumberNode(_token('number', filename, value))
+
+
+def bool(value: builtins.bool, filename: str) -> mparser.BooleanNode:
```

Comment:
Isn't this dangerous to have a function name that is a python keyword?  You appended an underscore to other such functions: `def or_()`.

Reply:
Likewise for Builder method.

Reply:
it's dangerous in as much that it would alias the builtin bool. But the intention is that you *don't* import as `from builder import ...` built just `import builder`, and then `builder.bool()` is perfectly safe to call because it's not aliasing the builtin. this is why the `builtins` module (as seen in the value) exists. I think the ergonomics of not having the `_` outweights the very small cost of having to use the `builtin` module inside of the `builder` module.

There is no cost to having an instance method called `bool`, since that has to be called with either `self` or a reference to the instance, ie `builder.bool()`

`or` is different because it's a keyword, and you *cannot* alias keywords. just like `assert` or `for`

Reply:
vscode highlight it wrong, that's how I noticed. But you're right, `bool()` is not a keyword, unlike `or`, that makes sense.